### PR TITLE
Fix validation in unpack phase of P2P shuffle

### PR DESCRIPTION
**Pull request by hendrikmakait**
This PR fixes the validation logic and canonicalizes some code.

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/shuffle/_core.py [hunk 1]
```diff
@@ -208,15 +208,33 @@ def _get_assigned_worker(self, i: _T_partition_id) -> str:
...
+    ) -> int:
+        """Add an input partition to the shuffle run"""
+
     async def get_output_partition(
         self, partition_id: _T_partition_id, key: str, **kwargs: Any
+    ) -> _T_partition_type:
+        self.raise_if_closed()
+        await self._ensure_output_worker(partition_id, key)
+        if not self.transferred:
+            raise RuntimeError("`get_output_partition` called before barrier task")
```
Comment (hendrikmakait) [inline distributed/shuffle/_core.py:231]: @fjetter: Switching these two checks around fixed the problem @pholf saw in testing.

* [ ] Add a deterministic test

Comment (hendrikmakait) [inline distributed/shuffle/_core.py:231]: @fjetter: Switching these two checks around fixed the problem @phofl saw in testing.

* [ ] Add a deterministic test

File: distributed/shuffle/tests/test_shuffle.py [hunk 2]
```diff
@@ -2207,3 +2207,65 @@ def test_sort_values_p2p_with_existing_divisions(client):
...
+    await invoke_annotation_chaos(1.0, c)
+
+    expected = pd.DataFrame({"a": list(range(10))})
+    ddf = dd.from_pandas(expected, npartitions=2)
+
+    # B should not participate in the initial shuffle phases, so place all outputs on A
+    def mock_get_worker_for_range_sharding(
+        output_partition: int, workers: list[str], npartitions: int
+    ) -> str:
+        return a.address
```
Comment (fjetter) [inline distributed/shuffle/tests/test_shuffle.py:2245]: This could also be established by starting a new worker after the shuffle started, couldn't it?

**Review by fjetter — APPROVED** [review]

File: distributed/shuffle/tests/test_shuffle.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/shuffle/tests/test_shuffle.py:2245]: Good point, that should also work. I'll see if I can simplify the test with that.

Comment (hendrikmakait) [reply distributed/shuffle/tests/test_shuffle.py]: I've simplified the test a bit.

File: distributed/shuffle/tests/test_shuffle.py [hunk 3]
```diff
@@ -2207,3 +2207,47 @@ def test_sort_values_p2p_with_existing_divisions(client):
...
+    shuffleA = get_shuffle_run_from_worker(shuffle_id, a)
+    await shuffleA.in_barrier.wait()
+
+    async with Worker(s.address) as b:
+        # Restrict an unpack task to B so that the previously non-participating
+        # worker takes part in the unpack phase
+        for key in s.tasks:
+            if key_split(key) == "shuffle_p2p":
+                s.set_restrictions({key: {b.address}})
+                break
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_shuffle.py:2249]: I stick with this to make this test robust to changes in the scheduling logic.

**Pull request merged**